language/stdlib/modules/validator_config_2.move

```diff
@@ -0,0 +1,111 @@
+    // The entity that can change the config is called a "validator operator".
+    // The owner of this resource can enable delegation and set delegated address.
+    // The owner can also disable delegation at any time.
+    resource struct T {
+        config: Config,
+        delegation_enabled: bool,
+        delegated_address: address,
+    }
+
+    // TODO(valerini): add events here
+
+    // Returns true if addr has a published ValidatorConfig::T resource
+    public fun has(addr: address): bool { exists<T>(addr) }
+
+    // Getters for Config
+    public fun get_config(addr: address): Config acquires T { *&borrow_global<T>(addr).config }
+    public fun get_consensus_pubkey(config_ref: &Config): vector<u8> { *&config_ref.consensus_pubkey }
+
+    // Returns the address of the entity who is now in charge of managing the config.
+    public fun get_validator_operator_address(addr: address): address acquires T {
+        let t_ref = borrow_global<T>(addr);
+        if (t_ref.delegation_enabled) {
+            return *&t_ref.delegated_address
+        } else {
+            return *&addr
+        }
+    }
+
+    // Register the transaction sender as a candidate validator by creating a ValidatorConfig
+    // resource under their account. Note that only one such resource can be instantiated under an account.
+    public fun register_candidate_validator(
+        consensus_pubkey: vector<u8>,) {
+
+        move_to_sender<T>(
+            T {
+                config: Config {
+                    consensus_pubkey: consensus_pubkey,
+                },
+                delegation_enabled: false,
+                delegated_address: 0x0
+            }
+        );
+    }
+
+    // If the sender of the transaction has a ValidatorConfig::T resource,
+    // this function will delegate management of the ValidatorConfig::T resource to a delegated_address.
+    public fun set_delegated_address(delegated_address: address) acquires T {
+        Transaction::assert(LibraAccount::exists(delegated_address), 5);
+        let t_ref = borrow_global_mut<T>(Transaction::sender());
+        *(&mut t_ref.delegation_enabled) = true;
+        *(&mut t_ref.delegated_address) = delegated_address;
+    }
+
+    // If the sender of the transaction has a ValidatorConfig::T resource,
+    // this function will revoke management capabilities from the delegated_address account.
+    public fun remove_delegated_address() acquires T {
+        let t_ref = borrow_global_mut<T>(Transaction::sender());
+        *(&mut t_ref.delegation_enabled) = false;
+        *(&mut t_ref.delegated_address) = 0x0;
+    }
+
+    // Rotate validator's config invoked by delegate account.
+    // Here validator_address - is the address of the account of the validator whose consensus_pubkey is going to be rotated.
+    public fun rotate_consensus_pubkey_by_delegate(validator_address: address, new_consensus_pubkey: vector<u8>, proof: vector<u8>) acquires T {
```

Comment:
did we wanna merge this with rotate_consensus_pubkey?

Reply:
We can merge, but then if the validator is the sender, it will have to pass in its own address, which is redundant as it can be obtained from `Transaction::sender()`.

Reply:
I unified the functions, while making the one with a shorter signature call into the one with a longer signature. Let me know if this looks better now!